iceoryx_hoofs/test/moduletests/test_static_lifetime_guard.cpp

```diff
@@ -0,0 +1,224 @@
+
+    // g_instance was constructed and the instance still exists
+    EXPECT_EQ(g_instance.id, 1);
+    EXPECT_EQ(Guard::count(), 1);
+    EXPECT_EQ(Foo::ctorCalled, 1);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, setCountWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1db790f9-d49e-44b2-b7e9-af50dd6a7d67");
+
+    auto oldCount = Guard::setCount(73);
+    EXPECT_EQ(Guard::count(), 73);
+    EXPECT_EQ(oldCount, 1);
+
+    Guard::setCount(oldCount);
+    EXPECT_EQ(Guard::count(), oldCount);
+}
+
+TEST_F(StaticLifetimeGuard_test, guardPreventsDestruction)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "5a8c5953-f2d7-4539-89ba-b4686bbb6319");
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+    EXPECT_EQ(g_instance.id, 1);
+    {
+        Guard guard;
+        EXPECT_EQ(Guard::count(), 2);
+        auto& instance = Guard::instance();
+
+        EXPECT_EQ(Foo::ctorCalled, 0);
+        EXPECT_EQ(Foo::dtorCalled, 0);
+
+        // still the same instance as g_instance
+        EXPECT_EQ(instance.id, 1);
+        EXPECT_EQ(&instance, &g_instance);
+    }
+
+    // the implicit guard of g_instance prevents destruction
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+    EXPECT_EQ(g_instance.id, 1);
+}
+
+TEST_F(StaticLifetimeGuard_test, copyIncreasesLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6ab6396d-7c63-4626-92ed-c7f3ea67bbf1");
+    Guard guard;
+    {
+        EXPECT_EQ(Guard::count(), 2);
+        // NOLINTJUSTIFICATION ctor and dtor side effects are tested
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        Guard copy(guard);
+        EXPECT_EQ(Guard::count(), 3);
+    }
+    EXPECT_EQ(Guard::count(), 2);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, moveIncreasesLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32a2fdbf-cb02-408c-99a3-373aa66b2764");
+    Guard guard;
+    {
+        EXPECT_EQ(Guard::count(), 2);
+        // NOLINTJUSTIFICATION ctor and dtor side effects are tested
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        Guard movedGuard(std::move(guard));
+        EXPECT_EQ(Guard::count(), 3);
+    }
+    EXPECT_EQ(Guard::count(), 2);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, assignmentDoesNotChangeLifetimeCount)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1c04ac75-d47a-44da-b8dc-6f567a53d3fc");
+    Guard guard1;
+    Guard guard2;
+
+    EXPECT_EQ(Guard::count(), 3);
+    guard1 = guard2;
+    EXPECT_EQ(Guard::count(), 3);
+    guard1 = std::move(guard2);
+    EXPECT_EQ(Guard::count(), 3);
+
+    EXPECT_EQ(Foo::ctorCalled, 0);
+    EXPECT_EQ(Foo::dtorCalled, 0);
+}
+
+TEST_F(StaticLifetimeGuard_test, destructionAtZeroCountWorks)
```

Comment:
This test cannot be run separately. Please make sure the `sut` is in a valid state at the end of the tests, e.g.
```cpp
Guard g;
g.instance();
g.setCount(2);
```

Reply:
Yes, this is a problem when static data is in play. I will improve this as far as possible to allow running the tests on their own (most of them should work already, I spend some effort before).

Reply:
I think in this specific case I got an error due to number of ctor calls